### PR TITLE
Refactor polymorphic branching

### PR DESCRIPTION
**Pull request by devmanhinton**
@yaymukund @bglusman 

Good refactor of collection reactions. Specifically their parsing.

There is more code then before and longer variable names. Longer code = bad. But I'm writing in a trade-off. Less succinct code for much clearer code.

Please let me know what you think.

### CODE REVIEW

File: lib/reactions/collection.js [hunk 1]
```diff
@@ -88,25 +88,44 @@ var CollectionReaction = Reaction.extend({
   }
 }, {
   selector: "[data-each]",
-    parse: function(el, state) {
```
Comment (bglusman) [inline lib/reactions/collection.js:91]: I guess the indentation here was just messed up before?

**Comment by yaymukund** [comment]
What's a `nonPolymorphicEnding`? How is this code better?

File: lib/reactions/collection.js (hunk 1 above)
Comment (devmanhinton) [inline lib/reactions/collection.js:91]: Yes. It was.

**Comment by devmanhinton** [comment]
@yaymukund 

This code is better because

-it clearly branches polymorphic and nonPolymorphic template parsing into two distinct parts
-it makes template lookUp clearer (before nonPolymorphic and Polymorphic named their templates
   lastItemOnPathStack + "default", when there was no whenValue.) nonPolymorphicEnding is the ending EndDash now adds onto templates if it is not polymorphic iteration
-Naming is better, templates is now childrensTemplates (which is a more accurate name)
-Better decomposition, related to first item, but before parse was doing several things without calling sub-functions which === bad decomp.

File: lib/reactions/collection.js [hunk 2]
```diff
@@ -88,25 +88,44 @@ var CollectionReaction = Reaction.extend({
...
+  },
+
+  parseDefault: function($child, state) {
+    var name = _.last(state.pathStack) + this.nonPolymorphicEnding;
+    this.registerAndRemoveChild($child, name, this.nonPolymorphicEnding);
+  },
+
+  registerAndRemoveChild: function($child, name, condition) {
+    EndDash.registerTemplate(name, $child);
+    this.childrensTemplates[condition] = EndDash.getTemplateClass(name);
```
Comment (yaymukund) [inline lib/reactions/collection.js]: I think internally, we should be using methods on `EndDash.TemplateStore` directly so dependencies are clearer.

If it's too verbose, you can also alias `templateStore` at the top of the file:

``` js
var ...
    EndDash = require('../end-dash'),
    templateStore = EndDash.templateStore;
```

Comment (devmanhinton) [inline lib/reactions/collection.js]: No. I agree with you.

For now, I'd prefer it to verbose and clear.

Let me change this.

**Comment by bglusman** [comment]
can't merge now, probably semi colins

**Comment by yaymukund** [comment]
The more I read the polymorphic code, the more I feel like we should just call this `switch-case` instead of `polymorphic`. I feel like that would be a lot clearer. The fact that we're branching on an attribute of the model is only incidental— for example, we branch on `call.status` but that isn't really a polymorphic attribute.

Also, maybe we could move it to its own reaction. Is there any reason for `polymorphic`(a.k.a. `switch-case`) to be in the same reaction as the collection stuff?

**Comment by tobowers** [comment]
:+1: on what mukund said... It's true that it's not necessarily polymorphic, but it is always a switch.  What about:

``` html
<div class="switchStatus- call-">
  <p class="whenHangup-">
    You done hungup
  </p>
</div>
```

**Comment by bglusman** [comment]
yeah, I like calling it a switch in some version too, particularly now we support an `else` case (and making that into `elseWhen-` or `else-` or something instead of just no class would be nice)

**Comment by yaymukund** [comment]
Errr, not sure I understand, @amicus-jenkins. Do you mean "Can one of the members of the end-dash team approve/merge this PR?"

While the old code wasn't great, I don't think this is much clearer either, to be honest.

**Comment by bglusman** [comment]
please test is what jenkins is looking for mukund

On Thursday, November 14, 2013, Mukund Lakshman wrote:

> Errr, not sure I understand, @amicus-jenkinshttps://github.com/amicus-jenkins.
> Do you mean "Can one of the members of the end-dash team approve/merge this
> PR?"
> 
> While the old code wasn't great, I don't think this is much clearer
> either, to be honest.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/Amicus/end-dash/pull/99#issuecomment-28487879
> .

## 

Brian Glusman - brian@amicushq.com
215-460-9585 - _cell_
_Twitter_: @amicushq

**Comment by yaymukund** [comment]
> please test is what jenkins is looking for mukund

Huh?

Am I supposed to type "please test?" Are you asking me to actually run the tests myself? Why? I don't think it should be merged even if it passes tests..

**Comment by yaymukund** [comment]
Oh, it's an automated message by [this plugin](https://wiki.jenkins-ci.org/display/JENKINS/GitHub+pull+request+builder+plugin)... yeah I'm going to ignore it.

**Comment by devmanhinton** [comment]
lol. I am going to close this for now.

**Pull request closed without merging**